packages/curve-ui-kit/src/shared/ui/ButtonMenu.tsx

```diff
@@ -0,0 +1,156 @@
+  executing?: boolean
+  /** Callback fired when the primary button is clicked */
+  onPrimary: () => void
+  /** Callback fired when a dropdown option is selected */
+  onOption?: (option: T) => void
+  /** Callback fired when the dropdown should open */
+  onOpen?: () => void
+  /** Callback fired when the dropdown should close */
+  onClose?: () => void
+}
+
+/**
+ * A split button component that combines a primary action button with an optional dropdown menu.
+ * The dropdown opens upward from the right side and contains additional action options.
+ *
+ * Menu implementation based on https://mui.com/material-ui/react-menu/#basic-menu
+ */
+export const ButtonMenu = <T extends string>({
+  primary,
+  options = [],
+  open,
+  disabled = false,
+  executing = false,
+  onPrimary,
+  onOption,
+  onOpen,
+  onClose,
+}: Props<T>) => {
+  const anchorEl = useRef<HTMLDivElement>(null)
+
+  return (
+    <Stack ref={anchorEl} direction="row" gap={'1px'}>
+      <Button color="primary" disabled={disabled || executing} sx={{ flexGrow: 1 }} onClick={onPrimary}>
+        {primary}
+      </Button>
+
+      {options.length > 0 && (
+        <Button
+          color="primary"
+          disabled={disabled || executing}
+          sx={{ width: ButtonSize.md, height: ButtonSize.md, minWidth: 'unset' }}
+          onClick={onOpen}
+        >
+          {executing ? <Spinner /> : <ChevronDownIcon sx={{ width: IconSize.lg, height: IconSize.lg }} />}
+        </Button>
+      )}
+
+      {options.length > 0 && (
+        <Menu
+          anchorEl={anchorEl.current}
+          open={open}
+          onClose={onClose}
+          // Modify anchor and transform to open upwards from the right
+          anchorOrigin={{
+            vertical: 'top',
+            horizontal: 'right',
+          }}
+          transformOrigin={{
+            vertical: 'bottom',
+            horizontal: 'right',
+          }}
+          slotProps={{
+            paper: {
+              sx: {
+                /**
+                 * Positions the menu with a small gap above the button component.
+                 *
+                 * - handleBreakpoints requires Responsive type values, preventing direct interpolation
+                 * - Uses CSS custom property to store spacing value for string interpolation
+                 * - !important is required to override MUI's internal high-specificity styles
+                 */
+                ...handleBreakpoints({
+                  '--options-gap': Spacing.sm,
+                  transform: `translateY(calc(-1 * var(--options-gap))) !important`,
+                }),
+
+                /**
+                 * Sets the menu width to match the anchor element (the button stack).
+                 * By default, MUI Menu auto-sizes to fit content, but we want consistent
+                 * width alignment with the trigger buttons (see Figma).
+                 */
+                width: anchorEl.current?.clientWidth,
```

Comment:
Using a ref is not going to rerender at the right time, you might want to `useResizeObserver`

Reply:
Ah right, it didn't resize *at all*, but it does now with the observer.